53 LDAP query bundle metadata

kevinleeMW opened the pull request:
Changes to allow LDAP query on bundle properties

Requirements

R1. Use LDAP queries against bundle/framework meta-data.
R2. Allow customization of LDAP search strategies against meta-data without impacting existing client code. 
R3. Make it easy and intuitive for C++ developers to query meta-data.
R4. LDAP query performance for full string match

jeffdiclemente commented:
What are the requirements for this feature? 
Knowing the requirements allows code reviewers to give more meaningful feedback.

jeffdiclemente commented on core/src/util/usProperties.cpp:
```diff
@@ -0,0 +1,130 @@
+    {
+        if (Find(iter->first) > -1)
+        {
+            std::string msg = "BundleProperties object contains case variants of the key: ";
```
This message is inconsistent with the one on line 67.

jeffdiclemente commented on core/test/usLDAPQueryTest.cpp:
```diff
@@ -0,0 +1,123 @@
+        if (bundle->GetBundleId() != 0)
+        {
+            bundle->Start();
+            //std::cout << bundle->GetProperty("bundle.name").ToString() << std::endl;
```
I'm assuming this no longer serves a purpose? If so, I'd suggest removing this commented code.

jeffdiclemente commented on core/test/usTestDriverActivator.cpp:
```diff
@@ -57,6 +57,7 @@ US_IMPORT_BUNDLE(CppMicroServices)
 US_IMPORT_BUNDLE(TestBundleA)
 US_IMPORT_BUNDLE(TestBundleA2)
 US_IMPORT_BUNDLE(TestBundleB)
+US_IMPORT_BUNDLE(TestBundleLQ_1)
```
Shouldn't all the other "TestBundleLQ_" bundles be listed here as statically linked bundles as well?

jeffdiclemente commented on core/src/util/usLDAPFilter.cpp:
```diff
@@ -0,0 +1,131 @@
+    
+bool LDAPFilter::Match(const Bundle& bundle) const
+{
+  return d->ldapExpr.Evaluate(Properties(bundle.d->bundleManifest.GetProperties()), true);
```
Why not call ```bundle.d->bundleManifest.GetProperties()``` instead? 
Doing this also eliminates adding LDAPFilter as a friend class to ```Bundle```.

karthikreddy09 replied:
+1. Bundle should have a GetProperties public method instead

jeffdiclemente commented on core/src/util/usLDAPExpr.cpp:
```diff
@@ -0,0 +1,852 @@
+/*=============================================================================
```
as we discussed offline, see if you can make git recognize this as a move of a file instead of a whole new file. If that is impossible, please summarize what areas have changed for code review. Thanks.

kevinleeMW replied:
I searched online, and found that it seems not possible with github UI. 
Summary of changes:
- use usProperties class instead of usServicePropertiesImpl class.

saschazelzer replied:
Some files like `usLDAPExpr.cpp` are still present in their original location (core/src/service). I guess this is the reason why the changes are not schown as a move. Could you remove the files, squash the commits and push-force the branch again?

jeffdiclemente commented on core/src/util/usLDAPExpr_p.h:
```diff
@@ -0,0 +1,180 @@
+/*=============================================================================
```
as we discussed offline, see if you can make git recognize this as a move of a file instead of a whole new file. If that is impossible, please summarize what areas have changed for code review. Thanks.

kevinleeMW replied:
I searched online, and found that it seems not possible with github UI. 
Summary of changes:
- use usProperties class instead of usServicePropertiesImpl class.

jeffdiclemente commented on core/src/util/usLDAPFilter.cpp:
```diff
@@ -0,0 +1,131 @@
+/*=============================================================================
```
as we discussed offline, see if you can make git recognize this as a move of a file instead of a whole new file. If that is impossible, please summarize what areas have changed for code review. Thanks.

kevinleeMW replied:
I searched online, and found that it seems not possible with github UI. 
Summary of changes:
- use usProperties class instead of usServicePropertiesImpl class.
- Added "Match(const Bundle& bundle)" method to enable LDAP query on bundle properties

jeffdiclemente commented on core/src/util/usProperties_p.h:
```diff
@@ -0,0 +1,57 @@
+  std::vector<std::string> keys;
+  std::vector<Any> values;
+
+  static Any emptyAny;
```
```static Any emptyAny``` should be ```const``` as it doesn't appear to change throughout its lifetime.

jeffdiclemente commented on core/src/util/usProperties_p.h:
```diff
@@ -0,0 +1,57 @@
+#ifndef USPROPERTIES_P_H
+#define USPROPERTIES_P_H
+
+#include "usServiceProperties.h"
```
headers should be self sufficient (i.e. they include everything they need for their declaration). See #66 for reasons why.

You should add ```<vector>```, ```<map>```, ```<string>``` and ```Any``` to satisfy the declaration.

jeffdiclemente commented on core/src/util/usProperties.cpp:
```diff
@@ -0,0 +1,130 @@
+    }
+    keys.push_back(iter->first);
+    values.push_back(iter->second);
+  }
```
There looks to be common code shared between the two constructors. I'd suggest refactoring to hoist the common code into one place. This makes maintenance a little easier.

jeffdiclemente commented on core/src/util/usProperties.cpp:
```diff
@@ -0,0 +1,130 @@
+
+const Any& Properties::Value(int index) const
+{
+  if (index < 0 || static_cast<std::size_t>(index) >= values.size())
```
even for one line if/else statements, always using braces. This cuts down on inadvertently introducing failures.

saschazelzer replied:
This is my old code, where I used one-line if statements without braces. Feel free to to add them though.

jeffdiclemente commented on core/test/usLDAPQueryTest.cpp:
```diff
@@ -0,0 +1,123 @@
+
+using namespace us;
+
+void TestLDAPQueryBundleMetaData(BundleContext* frameworkCtx)
```
If usTestBundleLQ_2 through usTestBundleLQ_5 do not export services and only contain meta-data, you shouldn't need to create dummy .cpp files for them. You should be able to generate "meta-data only" bundles. This will cut down on the number of files to maintain.

jeffdiclemente commented on core/test/usLDAPQueryTest.cpp:
```diff
@@ -0,0 +1,123 @@
+    
+    FrameworkFactory factory;
+    
+    std::shared_ptr<Framework> framework = factory.NewFramework(std::map<std::string, std::string>());
```
this can be simplified by removing ```std::map<std::string, std::string>()``` since no configuration is required.

jeffdiclemente commented on core/test/usLDAPQueryTest.cpp:
```diff
@@ -0,0 +1,123 @@
+    std::string bundleName;
+    Any bundleTestProperty;
+    
+    for (std::shared_ptr<Bundle> bundle : bundles)
```
I'd suggest using a ```const auto&``` for the range declaration to leverage type deduction and maintain const correctness. The ```bundle``` object will not be modified in the loop and ```LDAPFilter::Match(const Bundle& )``` expects a ```const&```.

jeffdiclemente commented on core/test/usLDAPQueryTest.cpp:
```diff
@@ -0,0 +1,123 @@
+    int actualMatchCount = 0;
+    int expectedMatchCount = 1;
+
+    for (std::shared_ptr<Bundle> bundle : bundles)
```
This loop appears to be unnecessary. Correct me if I'm wrong; the point of this test looks to be to test that query service properties work. 
If that's that case, its inefficient to iterate over all the bundles which don't export a service (unless there is a requirement I'm missing). The same result can be reached by using ```BundleContext::GetBundle(const std::string& name)```, passing in the name of the bundle exporting the service.

jeffdiclemente commented on core/test/usLDAPQueryTest.cpp:
```diff
@@ -0,0 +1,123 @@
+        }
+    }
+    
+    std::string test_description = "Test LDAP query of service properties (expected # of matches: " + std::to_string(expectedMatchCount) + "  actual # of matches: " + std::to_string(actualMatchCount)+ ")";
```
nit-pick: spacing of the ```+``` in ```std::to_string(actualMatchCount)+ ")";```

jeffdiclemente commented on core/test/usLDAPQueryTest.cpp:
```diff
@@ -0,0 +1,123 @@
+    framework->Stop();
+    
+    US_TEST_END()
+}
```
nit-pick: no newline at end of the file.

jeffdiclemente commented on core/include/usBundle.h:
```diff
@@ -398,6 +409,7 @@ class US_Core_EXPORT Bundle : public std::enable_shared_from_this<Bundle>
   friend class BundleRegistry;
   friend class ServiceReferencePrivate;
   friend class Framework;
+  friend class LDAPFilter;
```
can be removed if ```LDAPFilter::Match(...)``` uses ```Bundle::GetProperties()```

karthikreddy09 commented on core/test/usLDAPQueryTest.cpp:
```diff
@@ -0,0 +1,123 @@
+    InstallTestBundle(frameworkCtx, "TestBundleLQ_2");
+    InstallTestBundle(frameworkCtx, "TestBundleLQ_3");
+    InstallTestBundle(frameworkCtx, "TestBundleLQ_4");
+    InstallTestBundle(frameworkCtx, "TestBundleLQ_5");
```
Why 5 bundles ? Consider adding the absolute minimum number of bundles required to satisfy the test cases

karthikreddy09 commented on core/include/usBundle.h:
```diff
@@ -244,6 +245,16 @@ class US_Core_EXPORT Bundle : public std::enable_shared_from_this<Bundle>
+   *
+   * @sa \ref MicroServices_BundleProperties
+   */
+  std::map<std::string, Any> GetProperties() const;
```
Any reason why we use an ordered map here and an unordered map for service properties ?

kevinleeMW replied:
I will address this issue in the second iteration of this task (incl. optimization, etc.)

saschazelzer replied:
We could probably rename the `ServiceProperties` typedef (a std::unordered_map) to `Dictionary` and use this consistently.

karthikreddy09 commented on core/test/usLDAPQueryTest.cpp:
```diff
@@ -0,0 +1,123 @@
+    for (std::shared_ptr<Bundle> bundle : bundles)
+    {
+        
+        if (bundle->GetBundleId() != 0)
```
spacing

karthikreddy09 commented on core/src/CMakeLists.txt:
```diff
@@ -65,13 +61,19 @@ set(_srcs
   util/usWaitCondition_p.h
 
   service/usServiceHooks_p.h
+  service/usServiceListenerEntry_p.h
```
Is this change related to LDAP issue ? if not i'd suggest creating another branch for this issue. I believe this falls under issue#67

kevinleeMW replied:
These changes were made, because some of the files were moved; and made additional changes as I realized that some of the files were missing from this list.

karthikreddy09 commented on core/src/util/usLDAPExpr.cpp:
```diff
@@ -0,0 +1,852 @@
+  if ((d->m_operator & SIMPLE) != 0)
+  {
+    // try case sensitive match first
+    int index = p.FindCaseSensitive(d->m_attrName);
```
This is wierd ... why do attempt a casesentive match at all if matchCase is false.

saschazelzer replied:
This was optimized for the general case were it is assumed that people use filters on entities that contain the attribute names as property keys with the same case and the filter compares property values most of the time. Case mismatches are assumed to be typos or "laziness".

jeffdiclemente commented on core/test/usLDAPQueryTest.cpp:
```diff
@@ -0,0 +1,102 @@
+    
+  BundleContext* frameworkCtx = framework->GetBundleContext();
+    
+  InstallTestBundle(frameworkCtx, "TestBundleLQ");
```
```InstallTestBundle``` returns the ```std::shared_ptr<Bundle>``` it installed. This makes the call to ```frameworkCtx->GetBundle``` on the next line superfluous.

jeffdiclemente commented on core/test/usLDAPQueryTest.cpp:
```diff
@@ -0,0 +1,102 @@
+  US_TEST_BEGIN("LDAPQueryTest");
+    
+  FrameworkFactory factory;
+  std::shared_ptr<Framework> framework = factory.NewFramework();
```
prefer using ```auto```

jeffdiclemente commented on core/test/usLDAPQueryTest.cpp:
```diff
@@ -0,0 +1,102 @@
+  BundleContext* frameworkCtx = framework->GetBundleContext();
+    
+  InstallTestBundle(frameworkCtx, "TestBundleLQ");
+  const auto& bundle = frameworkCtx->GetBundle("TestBundleLQ");
```
why ```const auto&```?

jeffdiclemente commented on core/test/usLDAPQueryTest.cpp:
```diff
@@ -0,0 +1,102 @@
+  std::shared_ptr<Framework> framework = factory.NewFramework();
+  framework->Start();
+    
+  BundleContext* frameworkCtx = framework->GetBundleContext();
```
You could remove the variable and place ```framework->GetBundleContext()``` inside ```InstallTestBundle``` (assuming ```frameworkCtx->GetBundle``` is removed).

jeffdiclemente commented on core/test/usLDAPQueryTest.cpp:
```diff
@@ -0,0 +1,102 @@
+  // Testing case-insensitive value (should fail)
+  US_TEST_CONDITION(!ldapValueMismatchCase.Match(sr), " Evaluating LDAP expr: " + ldapValueMismatchCase.ToString());
+
+  bundle->Stop();
```
What's the intended behavior of ```LDAPFilter::Match(ServiceReference)``` once the bundle has stopped (and the service reference is no longer valid)?
Does this behavior need to be tested?

kevinleeMW replied:
Quoting OSGi spec (Section 5.2.1):
"A ````ServiceReference```` object is valid only as long as the service is registered. However, its properties must remain available as long as the ````ServiceReference```` object exists."

From this, I think that ```LDAPFilter::Match(ServiceReference)```` should behave the same even after the bundle has stopped. 

Perhaps, an additional test condition can be added after the bundle stops to make sure that querying service properties via a ````ServiceReferenceBase```` object still works.

What do you guys think?

jeffdiclemente replied:
I think it sounds pretty clear that service properties can be queried after the service they reference has been unregistered (e.g. on bundle stop).

What happens if you duplicate the four ```US_TEST_CONDITION``` macros after the ```bundle->Stop();``` and re-run the test?

kevinleeMW replied:
The first of the four ````US_TEST_CONDITION```` macros will fail, because it checks to make sure that the ServiceReference object is valid, and it no longer is after the ````bundle->Stop();````. 

Including the second test macro (that does the querying and expects a match) after the ````bundle->Stop();```` seems sufficient, as it will be testing that the service properties are still available (for LDAP queries).

jeffdiclemente commented on core/test/usLDAPQueryTest.cpp:
```diff
@@ -0,0 +1,102 @@
+
+using namespace us;
+
+void TestLDAPFilterMatchBundle(const std::shared_ptr<Bundle> bundle)
```
use ```const std::shared_ptr<Bundle>& bundle``` instead. That way there is no reference counting (which is unnecessary in this case and expensive) at function entry and exit.

jeffdiclemente commented on core/test/usLDAPQueryTest.cpp:
```diff
@@ -0,0 +1,102 @@
+  US_TEST_CONDITION(!ldapValueMismatchCase.Match(*bundle), " Evaluating LDAP expr: " + ldapValueMismatchCase.ToString());
+}
+
+void TestLDAPFilterMatchServiceReferenceBase(std::shared_ptr<Bundle> bundle)
```
use ```const std::shared_ptr<Bundle>& bundle``` instead. That way there is no reference counting (which is unnecessary in this case and expensive) at function entry and exit.

jeffdiclemente commented on core/test/usLDAPQueryTest.cpp:
```diff
@@ -0,0 +1,102 @@
+
+#include <usBundle.h>
+#include <usLDAPFilter.h>
+#include <usLog.h>
```
is this needed? I don't see any invocations of the logging macros.

jeffdiclemente commented on core/test/usLDAPQueryTest.cpp:
```diff
@@ -0,0 +1,102 @@
+#include "usTestingMacros.h"
+#include "usProperties_p.h"
+#include "usServiceReferenceBasePrivate.h"
+
```
I don't believe you want to ```#include``` private headers here (```usProperties_p.h``` and ```usServiceReferenceBasePrivate.h```). The public headers should be sufficient.

kevinleeMW commented:
I went through and addressed all of the feedback from Jeff, Abhinay and Sascha. Some of the suggestions will be tracked in Issue #85 as a follow-up task.

Is there any other flags from the pull request, @jeffdiclemente @saschazelzer @karthikreddy09 ?

jeffdiclemente commented:
I have no additional flags.
Thanks for creating #85 to further optimize.

saschazelzer commented:
Looks good to me, +1 for merging.

karthikreddy09 commented:
Ship It !
